kover-gradle-plugin/src/main/kotlin/kotlinx/kover/gradle/plugin/appliers/ProjectApplier.kt

```diff
@@ -29,122 +32,113 @@ internal class ProjectApplier(private val project: Project) {
+        project.dependencies.add(
+            JVM_REPORTER_CONFIGURATION_NAME,
+            toolProvider.map { tool -> tool.jvmReporterExtraDependency })
+
+        val defaultApplier = DefaultVariantApplier(
+            project,
+            koverDependencies,
+            reporterClasspath,
+            toolProvider
+        )
 
-        val instrData = collectInstrumentationData(tool, agentClasspath)
-        val locator = CompilationsLocatorFactory.get(project)
+        val instrData = collectInstrumentationData(toolProvider, agentClasspath)
 
-        val locate = locator.locate(projectExtension)
+        val listener = object : CompilationsListener {
+            override fun onJvmCompilation(kit: JvmCompilationKit) {
+                kit.tests.instrument(instrData)
+                defaultApplier.applyCompilationKit(kit)
+            }
 
-        val androidVariants = locate.android.associate {
-            it.buildVariant to project.createAndroidVariant(locate.kotlinPlugin, it, instrData, tool.variant)
-        }
+            override fun onAndroidCompilations(kits: List<AndroidVariantCompilationKit>) {
+                kits.forEach { kit ->
+                    kit.tests.instrument(instrData)
+                    val applier =
+                        AndroidVariantApplier(project, kit.buildVariant, koverDependencies, reporterClasspath, toolProvider)
 
-        val defaultVariant = createDefaultVariant(locate.kotlinPlugin, locate.jvm, instrData, tool.variant)
+                    val configs =
+                        reportExtension.namedReports[kit.buildVariant] ?: project.androidReports(kit.buildVariant)
 
-        reportExtension.default.merged.forEach { addedVariant ->
-            val androidVariant = androidVariants[addedVariant] ?: throw KoverIllegalConfigException("Android build variant '$addedVariant' was not found - it is impossible to add it to the default report")
+                    applier.applyConfig(configs, reportExtension.filters)
+                    applier.applyCompilationKit(kit)
 
-            defaultVariant.localArtifactGenerationTask.configure {
-                additionalArtifacts.from(androidVariant.localArtifact, androidVariant.dependentArtifactsConfiguration)
-                dependsOn(androidVariant.localArtifactGenerationTask, androidVariant.dependentArtifactsConfiguration)
+                    androidAppliers[kit.buildVariant] = applier
+                }
             }
-        }
-
-        ReportsApplier(defaultVariant, project, instrData.tool, reporterClasspath)
-            .createReports(reportExtension.default, reportExtension.filters)
 
-        androidVariants.forEach { (variantName, androidVariant) ->
-            val androidReport = reportExtension.namedReports[variantName] ?: project.objects.androidReports(variantName, project.layout)
+            override fun onFinalize() {
+                reportExtension.namedReports.keys.forEach { variantName ->
```

Comment:
This check should be run for regular JVM projects, in case the user accidentally configured the Android report.
This check always works, even if Kotlin plugin is not applied in the project (for the reason, see above).